### PR TITLE
hid: Implement SevenSixAxis and ConsoleSixAxisSensor

### PR DESCRIPTION
**Pull request by german77**
Implements SevenSixAxis and ConsoleSixAxisSensor. Allows to move the VR camera with motion.

This PR still needs more research and proper calibration of the motion sensors since they don't necessarily match with the console motion at the moment. But they are working well enough to be used in games that have this feature 

image

### CODE REVIEW

File: src/core/hle/service/hid/controllers/console_sixaxis.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 yuzu emulator team
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/console_sixaxis.cpp:1]: ditto with the ones below
```suggestion
// Copyright 2021 yuzu Emulator Project
```

File: src/core/hle/service/hid/controllers/console_sixaxis.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    // Called on InitializeSevenSixAxisSensor
+    void SetTransferMemoryPointer(u8* t_mem_1);
+
+    // Called on ResetSevenSixAxisSensorTimestamp
+    void ResetTimestamp();
+
+private:
+    struct SevenSixAxisState {
+        INSERT_PADDING_BYTES(0x10); // unused
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/console_sixaxis.h:40]:
```suggestion
        INSERT_PADDING_WORDS(4); // unused
```

File: src/core/hle/service/hid/controllers/console_sixaxis.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        s64_le sampling_number;
+        s64_le sampling_number2;
+        u64 unknown;
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/console_sixaxis.h:43]:
```suggestion
        s64_le sampling_number{};
        s64_le sampling_number2{};
        u64 unknown{};
```

File: src/core/hle/service/hid/controllers/console_sixaxis.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        Common::Vec3f accel{};
+        Common::Vec3f gyro{};
+        Common::Quaternion<f32> quaternion;
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/console_sixaxis.h:46]:
```suggestion
        Common::Quaternion<f32> quaternion{};
```

File: src/core/hle/service/hid/controllers/console_sixaxis.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    };
+    static_assert(sizeof(SevenSixAxisMemory) == 0xA70, "SevenSixAxisMemory is an invalid size");
+
+    struct ConsoleSharedMemory {
+        u64_le sampling_number;
+        bool is_seven_six_axis_sensor_at_rest;
+        f32 verticalization_error;
+        Common::Vec3f gyro_bias{};
+    };
+    static_assert(sizeof(ConsoleSharedMemory) == 0x20, "SharedMemory is an invalid size");
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/console_sixaxis.h:62]:
```suggestion
    static_assert(sizeof(ConsoleSharedMemory) == 0x20, "ConsoleSharedMemory is an invalid size");
```

File: src/core/hle/service/hid/controllers/console_sixaxis.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        CommonHeader header;
+        std::array<SevenSixAxisState, 0x21> sevensixaxis_states;
+    };
+    static_assert(sizeof(SevenSixAxisMemory) == 0xA70, "SevenSixAxisMemory is an invalid size");
+
+    struct ConsoleSharedMemory {
+        u64_le sampling_number;
+        bool is_seven_six_axis_sensor_at_rest;
+        f32 verticalization_error;
+        Common::Vec3f gyro_bias{};
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/console_sixaxis.h:60]: Default initialize all the members of these structs

File: src/core/hle/service/hid/hid.cpp [hunk 7]
```diff
@@ -1404,8 +1403,9 @@ void Hid::ActivateConsoleSixAxisSensor(Kernel::HLERequestContext& ctx) {
     IPC::RequestParser rp{ctx};
     const auto applet_resource_user_id{rp.Pop<u64>()};
 
-    LOG_WARNING(Service_HID, "(STUBBED) called, applet_resource_user_id={}",
-                applet_resource_user_id);
+    LOG_WARNING(Service_HID, "called, applet_resource_user_id={}", applet_resource_user_id);
+
+    applet_resource->ActivateController(HidController::ConsoleSixAxisSensor);
```
Comment (Morph1984) [inline src/core/hle/service/hid/hid.cpp:1408]: I'd place these above the log

File: src/core/hle/service/hid/hid.cpp [hunk 8]
```diff
@@ -1455,8 +1455,9 @@ void Hid::ActivateSevenSixAxisSensor(Kernel::HLERequestContext& ctx) {
...
```
Comment (Morph1984) [inline src/core/hle/service/hid/hid.cpp:1460]: ditto

File: src/core/hle/service/hid/hid.cpp [hunk 9]
```diff
@@ -1519,10 +1520,16 @@ void Hid::InitializeSevenSixAxisSensor(Kernel::HLERequestContext& ctx) {
...
+                "called, t_mem_1_handle=0x{:08X}, t_mem_2_handle=0x{:08X}, "
                 "applet_resource_user_id={}",
                 t_mem_1_handle, t_mem_2_handle, applet_resource_user_id);
 
+    // Activate console six axis controller
+    applet_resource->GetController<Controller_ConsoleSixAxis>(HidController::ConsoleSixAxisSensor)
+        .ActivateController();
+
+    applet_resource->GetController<Controller_ConsoleSixAxis>(HidController::ConsoleSixAxisSensor)
+        .SetTransferMemoryPointer(t_mem_1->GetPointer());
```
Comment (Morph1984) [inline src/core/hle/service/hid/hid.cpp:1532]: ditto

File: src/core/hle/service/hid/hid.cpp [hunk 10]
```diff
@@ -1542,8 +1549,10 @@ void Hid::ResetSevenSixAxisSensorTimestamp(Kernel::HLERequestContext& ctx) {
     IPC::RequestParser rp{ctx};
     const auto applet_resource_user_id{rp.Pop<u64>()};
 
-    LOG_WARNING(Service_HID, "(STUBBED) called, applet_resource_user_id={}",
-                applet_resource_user_id);
+    LOG_WARNING(Service_HID, "called, applet_resource_user_id={}", applet_resource_user_id);
+
+    applet_resource->GetController<Controller_ConsoleSixAxis>(HidController::ConsoleSixAxisSensor)
```
Comment (Morph1984) [inline src/core/hle/service/hid/hid.cpp:1554]: ditto

File: src/core/hle/service/hid/controllers/console_sixaxis.cpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    const auto& last_entry =
+        seven_six_axis.sevensixaxis_states[seven_six_axis.header.last_entry_index];
+    seven_six_axis.header.last_entry_index = (seven_six_axis.header.last_entry_index + 1) % 17;
+    auto& cur_entry = seven_six_axis.sevensixaxis_states[seven_six_axis.header.last_entry_index];
+
+    cur_entry.sampling_number = last_entry.sampling_number + 1;
+    cur_entry.sampling_number2 = cur_entry.sampling_number;
+
+    // Try to read sixaxis sensor states
+    MotionDevice motion_devices{};
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/console_sixaxis.cpp:41]:
```suggestion
    MotionDevice motion_device{};
```

File: src/core/hle/service/hid/controllers/console_sixaxis.cpp [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    const auto player = Settings::values.players.GetValue()[0];
+    std::transform(player.motions.begin() + Settings::NativeMotion::MOTION_HID_BEGIN,
+                   player.motions.begin() + Settings::NativeMotion::MOTION_HID_END, motions.begin(),
+                   Input::CreateDevice<Input::MotionDevice>);
+}
+
+void Controller_ConsoleSixAxis::SetTransferMemoryPointer(u8* t_mem_1) {
+    is_transfer_memory_set = true;
+    transfer_memory = t_mem_1;
+};
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/console_sixaxis.cpp:83]:
```suggestion
}
```

File: src/core/hle/service/hid/controllers/console_sixaxis.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    void OnRelease() override;
+
+    // When the controller is requesting an update for the shared memory
+    void OnUpdate(const Core::Timing::CoreTiming& core_timing, u8* data, size_t size) override;
+
+    // Called when input devices should be loaded
+    void OnLoadInputDevices() override;
+
+    // Called on InitializeSevenSixAxisSensor
+    void SetTransferMemoryPointer(u8* t_mem_1);
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/console_sixaxis.h:33]:
```suggestion
    void SetTransferMemoryPointer(u8* t_mem);
```

File: src/core/hle/service/hid/controllers/console_sixaxis.cpp [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+void Controller_ConsoleSixAxis::OnLoadInputDevices() {
+    const auto player = Settings::values.players.GetValue()[0];
+    std::transform(player.motions.begin() + Settings::NativeMotion::MOTION_HID_BEGIN,
+                   player.motions.begin() + Settings::NativeMotion::MOTION_HID_END, motions.begin(),
+                   Input::CreateDevice<Input::MotionDevice>);
+}
+
+void Controller_ConsoleSixAxis::SetTransferMemoryPointer(u8* t_mem_1) {
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/console_sixaxis.cpp:80]:
```suggestion
void Controller_ConsoleSixAxis::SetTransferMemoryPointer(u8* t_mem) {
```

File: src/core/hle/service/hid/controllers/console_sixaxis.h [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+        Common::Vec3f accel;
+        Common::Vec3f gyro;
+        Common::Vec3f rotation;
+        std::array<Common::Vec3f, 3> orientation;
+        Common::Quaternion<f32> quaternion;
+    };
+
+    using MotionArray =
+        std::array<std::unique_ptr<Input::MotionDevice>, Settings::NativeMotion::NUM_MOTIONS_HID>;
+    u8* transfer_memory;
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/console_sixaxis.h:74]: This should be initialized to nullptr here, for a a consistent constructed state.

**Review by bunnei — APPROVED** [review]

**Pull request merged**